### PR TITLE
Database: Add abstract database class

### PR DESCRIPTION
**Pull request by bergzand**
Paves the way for implementing alternative database back-ends.

There should be a better way to mock the individual database functions for a database back-end.

### CODE REVIEW

**Comment by aabadie** [comment]
I think this is half baked :)

What I see missing:
- a basic test for the base class (it's not covered as is)
- I would like to see a kind of factory function, like `database(db_type: str) -> Database` that returns an instance of the right database and that is use to instanciate the `db` attribute of the murdock class.
- the db_type above could be retrieved automatically from settings, with associated settings depending on the type of DB. The default would be mongodb ;)

With the 2 last points, it will super simple to switch from one DB type to the other as it would just be a matter of settings in the .env file.

I know I'm asking for much more work, let me know if you don't feel like doing that.

**Comment by bergzand** [comment]
> I think this is half baked :)

I didn't want to put more in this PR than what is strictly necessary :)


> a basic test for the base class (it's not covered as is)

Honest question, what should be in this test? The whole base class is abstract without any implementation. Or do you mean an empty test with the whole class as mock so that it is covered?

> I would like to see a kind of factory function, like `database(db_type: str) -> Database` that returns an instance of the right database and that is use to instantiate the `db` attribute of the murdock class.

Makes sense, I can add something like this.

> the db_type above could be retrieved automatically from settings, with associated settings depending on the type of DB. The default would be mongodb ;)

I have something like this in a follow up PR, I can move it to this PR.

**Comment by aabadie** [comment]
> Honest question, what should be in this test? The whole base class is abstract without any implementation. Or do you mean an empty test with the whole class as mock so that it is covered?

I'll comment inline because I don't think an explicit test is necessary: it's already tested via the concrete mongo db implementation.

File: murdock/database/__init__.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class Database(abc.ABC):
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @abc.abstractmethod
+    async def init(self):
+        pass
```
Comment (aabadie) [inline murdock/database/__init__.py:15]: I prefer the `...` notation :)
Also you can skip the coverage using a pragma comment (same comment in other abstract methods below).

```suggestion
        ...  # pragma: no cover
```

Comment (bergzand) [reply murdock/database/__init__.py]: I annotated all functions with docstring, that also removes the need for `pass` and `...`. Probably doesn't satisfy the coverage, but lets see.

**Comment by bergzand** [comment]
> I would like to see a kind of factory function, like `database(db_type: str) -> Database` that returns an instance of the right database and that is use to instanciate the `db` attribute of the murdock class.

Added one to directly instantiate a class based on the argument and one that uses the configuration from the environment variables

**Comment by aabadie** [comment]
Needs a rebase

**Review by aabadie — APPROVED** [review]

**Pull request merged**